base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -438,6 +438,87 @@ impl WalletBackend for WalletSqliteDatabase {
+                let mut secondary_key = WalletSecondaryEncryptionKey::from(SafeArray::default());
+                argon2::Argon2::new(argon2_params.algorithm, argon2_params.version, argon2_params.params)
+                    .hash_password_into(
+                        existing.reveal(),
+                        secondary_key_salt.as_bytes(),
+                        secondary_key.reveal_mut(),
+                    )
+                    .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
+
+                // Attempt to decrypt the encrypted main key
+                let main_key_cipher = XChaCha20Poly1305::new(Key::from_slice(secondary_key.reveal()));
+                let mut aad = main_key_domain.clone();
+                aad.push(version);
+
+                let main_key = WalletMainEncryptionKey::from(
+                    decrypt_bytes_integral_nonce(
+                        &main_key_cipher,
+                        aad,
+                        &from_hex(&encrypted_main_key)
+                            .map_err(|e| WalletStorageError::ConversionError(e.to_string()))?,
+                    )
+                    .map_err(|_| WalletStorageError::InvalidPassphrase)?,
+                );
+
+                // Derive a new secondary key from the new passphrase and a fresh salt
+                argon2_params = Argon2Parameters::from_version(None)?;
+                let new_secondary_key_salt = SaltString::generate(&mut OsRng);
+                let mut new_secondary_key = WalletSecondaryEncryptionKey::from(SafeArray::default());
+                argon2::Argon2::new(argon2_params.algorithm, argon2_params.version, argon2_params.params)
+                    .hash_password_into(
+                        new.reveal(),
+                        new_secondary_key_salt.as_bytes(),
+                        new_secondary_key.reveal_mut(),
+                    )
+                    .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
+
+                // Encrypt the main key with the new secondary key
+                let new_main_key_cipher = XChaCha20Poly1305::new(Key::from_slice(new_secondary_key.reveal()));
+                let mut new_aad = main_key_domain;
+                new_aad.push(argon2_params.id);
+                let new_encrypted_main_key = encrypt_bytes_integral_nonce(
+                    &new_main_key_cipher,
+                    new_aad,
+                    Hidden::hide(main_key.reveal().clone()),
+                )
+                .map_err(WalletStorageError::AeadError)?;
+
+                // Store the secondary key version, secondary key salt, and encrypted main key
+                WalletSettingSql::new(DbKey::SecondaryKeyVersion, argon2_params.id.to_string()).set(&conn)?;
+                WalletSettingSql::new(DbKey::SecondaryKeySalt, new_secondary_key_salt.to_string()).set(&conn)?;
+                WalletSettingSql::new(DbKey::EncryptedMainKey, new_encrypted_main_key.to_hex()).set(&conn)?;
```

Comment:
This makes 3 transactions. 
It should be 1 here, this is the crucial part we need to be atomic

Reply:
Yep! I'm testing code locally that does this.